### ISSUE TITLE
[BUG]: erdos_renyi_generator not working properly

### ISSUE THREAD

**Issue by lvjing421** [rapidsai/cugraph#4359]
### Version

v23.02.00

### Which installation method(s) does this occur on?

Conda

### Describe the bug.

The source is cugraph/cpp/src/generators/erdos_renyi_generator.cu, and the test is cugraph/cpp/tests/generators/erdos_renyi_test.cpp. When random graph generation probability is changed from 0.1 to 0.5, there is some test error, e.g., with er_test<int32_t>(size_t{10}, float{0.5}). Moreover, the strange thing is that almost all generated source-to-destination vertice pairs are 0-0s, meaning there is almost no edge between vertex 0 and other vertices. I wonder if the function is mature enough for usage or is it to be further developed? Thank you for your time.

### Code of Conduct

- [X] I agree to follow cuGraph's Code of Conduct
- [X] I have searched the [open bugs](https://github.com/rapidsai/cugraph/issues?q=is%3Aopen+is%3Aissue+label%3Abug) and have found no duplicates for this bug report

**Comment by ChuckHastings** [comment]
I can reproduce your error.  I will investigate.

This has not been used in our code, and it is not exposed in our python API.  Consequently, it is not regularly tested.  I would expect this to work, this is clearly a bug.  I will isolate this next week.

**Comment by ChuckHastings** [comment]
I was able to isolate and correct.  Once the linked PR is merged, you should be able to see the updated source code.

I notice that you were using a pretty old version of cugraph.  23.02 was released in February of last year.  We just released version 24.04.  This bug still exists in 24.04, my fix should appear in the source code once merged and in the nightly builds the following day.  It will be released in 24.06.

**Comment by lvjing421** [comment]
Hi Chuck,

Thank you for the quick solution. I have tried it with v23.02.00 - having not the opportunity to test with v24.04.00 or newer. The compiler says thrust::make_transform_output_iterator() is not copyable in thrust::copy_if() as its copy assignment operator is implicitly deleted. Is it possible to use other transform-like approach to work around?

With regards,
Jing

**Comment by ChuckHastings** [comment]
What version of CUDA are you using?

**Comment by ChuckHastings** [comment]
I would suggest trying out the 24.06 branch which now has this feature.  I have tested this with CUDA 12.x and CUDA 11.8 and it seems to work properly.  If you're using an older version of CUDA that might also be an issue.

If you continue to have problems please provide more details of your build environment (CUDA version, compiler versions, etc) and either create a new issue or reopen this one.

**Comment by lvjing421** [comment]
Hi, I am working with cudatoolkit=11.8 and nvcc_linux-64=11.8 as in https://github.com/rapidsai/cugraph/blob/v23.02.00/conda/environments/all_cuda-118_arch-x86_64.yaml. The compilation error is as follows, which states that in thrust::copy_if(), copy assignment operator of thrust::make_transform_output_iterator() is deleted - is it because the lamba expression argument of thrust::make_transform_output_iterator() has a deleted copy assignment operator as in https://en.cppreference.com/w/cpp/language/lambda, or thrust::make_transform_output_iterator() itself? 
---------------------------------------------------------------------------------------------------------
[  0%] Building CUDA object CMakeFiles/cugraph.dir/src/generators/erdos_renyi_generator.cu.o
/home/jinglv/miniconda3/envs/cugraph_dev/bin/nvcc -forward-unknown-to-host-compiler -DCUDA_API_PER_THREAD_DEFAULT_STREAM -DCUTLASS_NAMESPACE=raft_cutlass -DRAFT_DISTANCE_COMPILED -DRAFT_SYSTEM_LITTLE_ENDIAN=1 -DTHRUST_DEVICE_SYSTEM=THRUST_DEVICE_SYSTEM_CUDA -DTHRUST_HOST_SYSTEM=THRUST_HOST_SYSTEM_CPP -Dcugraph_EXPORTS --options-file CMakeFiles/cugraph.dir/includes_CUDA.rsp -O3 -DNDEBUG -std=c++17 "--generate-code=arch=compute_86,code=[sm_86]" -Xcompiler=-fPIC --expt-extended-lambda --expt-relaxed-constexpr -Werror=cross-execution-space-call -Wno-deprecated-declarations -Xptxas=--disable-warnings -Xcompiler=-Wall,-Wno-error=sign-compare,-Wno-error=unused-but-set-variable -Xfatbin=-compress-all -Xcompiler -pthread -MD -MT CMakeFiles/cugraph.dir/src/generators/erdos_renyi_generator.cu.o -MF CMakeFiles/cugraph.dir/src/generators/erdos_renyi_generator.cu.o.d -x cu -c /data/jinglv/bigdata/cugraph/cpp/src/generators/erdos_renyi_generator.cu -o CMakeFiles/cugraph.dir/src/generators/erdos_renyi_generator.cu.o
/home/jinglv/miniconda3/envs/cugraph_dev/include/rapids/thrust/system/cuda/detail/copy_if.h(796): error: function "thrust::transform_output_iterator<UnaryFunction, OutputIterator>::operator=(const thrust::transform_output_iterator<lambda [](size_t)->thrust::tuple<int, int, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>, thrust::zip_iterator<thrust::tuple<int32_t *, int32_t *, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>>> &) [with UnaryFunction=lambda [](size_t)->thrust::tuple<int, int, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>, OutputIterator=thrust::zip_iterator<thrust::tuple<int32_t *, int32_t *, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>>]" (declared implicitly) cannot be referenced -- it is a deleted function
          detected during:
            instantiation of "OutputIterator thrust::cuda_cub::copy_if(thrust::cuda_cub::execution_policy<Derived> &, InputIterator, InputIterator, OutputIterator, Predicate) [with Derived=thrust::detail::execute_with_allocator<rmm::mr::thrust_allocator<char>, thrust::cuda_cub::execute_on_stream_base>, InputIterator=thrust::counting_iterator<size_t, thrust::use_default, thrust::use_default, thrust::use_default>, OutputIterator=thrust::transform_output_iterator<lambda [](size_t)->thrust::tuple<int, int, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>, thrust::zip_iterator<thrust::tuple<int32_t *, int32_t *, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>>>, Predicate=lambda [](size_t)->__nv_bool]" 
/home/jinglv/miniconda3/envs/cugraph_dev/include/rapids/thrust/detail/copy_if.inl(41): here
            instantiation of "OutputIterator thrust::copy_if(const thrust::detail::execution_policy_base<DerivedPolicy> &, InputIterator, InputIterator, OutputIterator, Predicate) [with DerivedPolicy=thrust::detail::execute_with_allocator<rmm::mr::thrust_allocator<char>, thrust::cuda_cub::execute_on_stream_base>, InputIterator=thrust::counting_iterator<size_t, thrust::use_default, thrust::use_default, thrust::use_default>, OutputIterator=thrust::transform_output_iterator<lambda [](size_t)->thrust::tuple<int, int, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>, thrust::zip_iterator<thrust::tuple<int32_t *, int32_t *, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type, thrust::null_type>>>, Predicate=lambda [](size_t)->__nv_bool]" 
/data/jinglv/bigdata/cugraph/cpp/src/generators/erdos_renyi_generator.cu(73): here
            instantiation of "std::tuple<rmm::device_uvector<vertex_t>, rmm::device_uvector<vertex_t>> cugraph::generate_erdos_renyi_graph_edgelist_gnp(const raft::handle_t &, vertex_t, float, vertex_t, uint64_t) [with vertex_t=int32_t]" 
/data/jinglv/bigdata/cugraph/cpp/src/generators/erdos_renyi_generator.cu(89): here

**Comment by ChuckHastings** [comment]
Let me try and create a 23.02 environment for 11.8...

**Comment by ChuckHastings** [comment]
OK.  I can recreate your issue.  When I asserted earlier about testing with 11.8, I was actually building with a 12.2 compiler and executing on 11.8.

I definitely recommend moving forward to a newer version.  However, if you have to build with this version, I was able to get it compiling with a 23.02 build environment.  Here's a diff of the changes that worked for me:

```
31,42d30
< namespace {
<   template <typename vertex_t>
<   struct index_to_vertex_pair {
<     vertex_t num_vertices;
< 
<     thrust::tuple<vertex_t, vertex_t> __device__ operator()(size_t index) {
<       return thrust::make_tuple(static_cast<vertex_t>(index / num_vertices),
<                                 static_cast<vertex_t>(index % num_vertices));
<     }
<   };
< }
< 
58c46
<   auto generate_random_value = [seed] __device__(size_t index) {
---
>   auto generate_random_value = cuda::proclaim_return_type<float>([seed] __device__(size_t index) {
63c51
<   };
---
>   });
80c68,72
<                     index_to_vertex_pair<vertex_t>{num_vertices}),
---
>                     cuda::proclaim_return_type<thrust::tuple<vertex_t, vertex_t>>(
>                       [num_vertices] __device__(size_t index) {
>                         return thrust::make_tuple(static_cast<vertex_t>(index / num_vertices),
>                                                   static_cast<vertex_t>(index % num_vertices));
>                       })),
```


Essentially I moved the lambda out into a separate struct, and I eliminated the `cuda::proclaim_return_type` from the definition of `generate_random_value`.

But this should work properly in a 24.06 environment without changes now that the PR was merged.

**Comment by lvjing421** [comment]
It works, and thanks for the effort.
